lib/ast/AST.cpp

```diff
@@ -285,6 +286,512 @@ sptr<KOREPattern> KORECompositePattern::expandAliases(KOREDefinition *def) {
+      colors["PaleTurquoise"] = "\x1b[38;5;159m";
+      colors["DarkGoldenrod"] = "\x1b[38;5;136m";
+      colors["lime"] = "\x1b[38;5;154m";
+      colors["DeepPink"] = "\x1b[38;5;198m";
+      colors["MediumVioletRed"] = "\x1b[38;5;162m";
+      colors["orange"] = "\x1b[38;5;220m";
+      colors["HotPink"] = "\x1b[38;5;205m";
+      colors["LightYellow"] = "\x1b[38;5;230m";
+      colors["Navy"] = "\x1b[38;5;18m";
+      colors["SaddleBrown"] = "\x1b[38;5;94m";
+      colors["Coral"] = "\x1b[38;5;209m";
+      colors["SlateGrey"] = "\x1b[38;5;102m";
+      once = false;
+    }
+    append(out, colors[color]);
+  }
+}
+
+std::string enquote(std::string str) {
+  std::string result;
+  for (size_t i = 0; i < str.length(); ++i) {
+    char c = str[i];
+    switch(c) {
+    case '\\':
+      result.append("\\\\");
+      break;
+    case '"':
+      result.append("\\\"");
+      break;
+    case '\n':
+      result.append("\\n");
+      break;
+    case '\t':
+      result.append("\\t");
+      break;
+    case '\r':
+      result.append("\\r");
+      break;
+    case '\f':
+      result.append("\\f");
+      break;
+    default:
+      if ((unsigned char)c >= 32 && (unsigned char)c < 127) {
+        result.push_back(c);
+      } else {
+        char buf[3];
+        buf[2] = 0;
+        snprintf(buf, 3, "%02x", (unsigned char)c);
+        result.append("\\x");
+        result.append(buf);
+      }
+      break;
+    }
+  }
+  return result;
+}
+
+void KORECompositePattern::prettyPrint(std::ostream &out, PrettyPrintData const& data) const {
+  std::string name = getConstructor()->getName();
+  if (name == "\\dv") {
+    KORECompositeSort *s = dynamic_cast<KORECompositeSort *>(getConstructor()->getFormalArguments()[0].get()); 
+    bool hasHook = data.hook.count(s->getName());
+    auto str = dynamic_cast<KOREStringPattern *>(arguments[0].get());
+    if (hasHook) {
+      auto hook = data.hook.at(s->getName());
+      if (hook == "STRING.String") {
+        append(out, enquote(str->getContents()));
+      } else if (hook == "BYTES.Bytes") {
+        append(out, 'b');
+        append(out, enquote(str->getContents()));
+      } else {
+        append(out, str->getContents());
+      }
+    } else {
+      append(out, str->getContents());
+    }
+    return;
+  }
+  if (data.format.count(name)) {
+    auto format = data.format.at(name);
+    int localIndent = 0;
+    int localColor = 0;
+    for (int i = 0; i < format.length(); ++i) {
+      char c = format[i];
+      if (c == '%') {
+        char c2 = format[i+1];
+        ++i;
+        switch(c2) {
+          case 'n':
+            newline();
+            break;
+          case 'i':
+            indent++;
+            localIndent++;
+            break;
+          case 'd':
+            indent--;
+            localIndent--;
+            break;
+          case 'c':
+            if (data.colors.count(name)) {
+              color(out, data.colors.at(name)[localColor++]);
```

Comment:
I don't understand this. Why is `localColor` incremented here? Could it happen that it is incremented past the end of that vector?

Reply:
That would only happen if the user incorrectly specifies their `format` or `colors` attributes.